### ISSUE TITLE
DeepCABAC

### ISSUE THREAD

**Issue by jihunoh-neubla** [NeublaCorp/DeepCABAC#1]
- Paper: 
[S. Wiedemann et al., "DeepCABAC: A Universal Compression Algorithm for Deep Neural Networks," IEEE Journal of selected topics in signal processing, May 2020.](https://arxiv.org/pdf/1905.08318.pdf)

- Original github: 
https://github.com/fraunhoferhhi/DeepCABAC

- Summary:
A universal compression algorithm for DNNs that is based on applying Context-based Adaptive Binary Arithmetic Coder (CABAC) to the DNN parameters. 
DeepCABAC applies a novel quantization scheme that minimizes a rate-distortion function while simultaneously taking the impact of quantization to the DNN performance into account.
For instance, it is able to compress the VGG16 ImageNet model by x63.6 with no loss of accuracy.

- Purpose:
To see the feasibility of this compression technique if applying to our NPU

**Comment by jihunoh-neubla** [comment]
### **Compression Algorithm**

Background
- Lossless coding
  1. Huffman coding
  2. Arithmetic coding
  3. **Universal coding**
     - Characteristics
       - Universality: adapt to a wide range of different types of input distribution
       - Minmial redundancy: binary representations of minimal redundancy 
       - High efficiency: high throughput
     - CABAC (context-based adaptive binary arithmetic coding, adopted to H.265/HEVC)
       1. Binarization - predefine a series of binary decision
       2. Context-modeling - assigns a binary probability model to each bin, updated on-the-fly by the local statistics of data
       3. Arithmetic conding - employs an arithmetic coder
- Lossy coding
  - the quantizer Q is non-invertible
  1. Scalar Lloyd Algorithm
  2. CABAC-based RD-quantization
    - Given a set of quantization points and select CABAC as universal lossless code

- DeepCABAC
  - Overall process
    1. Extract the weight params of NN layer-by-layer in row-major order
    2. Select a value \beta which defines the set of quantization points
    3. subsequently quantizes the weight values by minimizing a weight rate-distortion function  
![image](https://user-images.githubusercontent.com/99005178/163303980-e3366a93-cdf4-46e5-add1-06b311e082ea.png)
    4. compress them by applying CABAC
    5. reconstruct the network and evaluates the prediction performance of the quantized model
    6. this process is repeated for a set of hyperparam \beta until the desired accuracy vs. size trade-off is achieved
  - Coder 
![image](https://user-images.githubusercontent.com/99005178/163313620-6152c6a3-9a4e-449c-82d2-7531f7b5c941.png)
  - Coder + Quantizer
    - Assignment: finding quantizer Q optimaaly assigns the quantization point to each weight param
    - Quant. points: finding the optimal quantization point value
      - quantizer has two configurable hyper params \beta = (\step_size, \lambda)
  

![image](https://user-images.githubusercontent.com/99005178/163292746-6e66e56b-0204-4786-8c08-9a4632ed23d5.png)
![image](https://user-images.githubusercontent.com/99005178/163303593-867e7da6-6f2d-4cf4-86a3-f19a48a58277.png)

**Comment by jihunoh-neubla** [comment]
### **Compression Performance**

- Model: Yolo v3 (input size = 608x608)
[/inshared/Model/vision/object_detection/yolo_v3/onnx/yolov3_d53_mstrain-608_273e_coco_20210518_115020-a2c3acb8_608.onnx
](/inshared/Model/vision/object_detection/yolo_v3/onnx/yolov3_d53_mstrain-608_273e_coco_20210518_115020-a2c3acb8_608.onnx)


- Method
  1. Our 8-bit uniform layer-wise asymmetric quantization + binary coding 
  2. Deepcabac's quantization (8-bit uniform layer-wise symmetric quantization) + binary coding

- Measure

  | Encoded </br> parameter size  (MB) | Mean diff </br> (fp32 - dequant. fp32) | Max diff </br> (fp32 - dequant. fp32) | mAP
-- | -- | -- | -- | --
original | 238 | - | - | 0.309
our quantization | 40 (x5.9) | 0.000806 | 0.041181 | 0.275
deepcabac quantization | 28 (x8.5) |  0.001353 | 11.312755 | -

**Comment by deumji-woo** [comment]
# Compression ratio result
1. Compression method
  * Didn't use DeepCABAC quantization.
  * Only use their "coder" part.
  * our quantization + "coder" of DeepCABAC
2. channel-wise symmetric quantization
  * model: /neublan-shared/Model/vision/object_detection/yolo_v3/onnx/yolov3_d53_mstrain-608_273e_coco_20210518_115020-a2c3acb8_608.q8_cwq_exc_concat_mul_add193+.onnx
  * compression ratio :  87%
    > [detail result](https://neubla.sharepoint.com/:x:/s/All/EX2b_uduiAxElCcR5HE6TisBZOrJ9QIEKhZCOvnVRKAmWQ?e=gsEzXN)
3. layer-wise asymmetric quantization
  * model: /neublan-shared/Model/vision/object_detection/yolo_v3/onnx/yolov3_d53_mstrain-608_273e_coco_20210518_115020-a2c3acb8_608.q8_lwq_exc_concat_mul_add193+.onnx
  * compression ratio :  68%
    > [detail result](https://neubla.sharepoint.com/:x:/s/All/Ea5IgP0sygxDrRKlu8-AzhkBPdC9xO2djisXP64VRZSoTQ?e=0zpGCD)

**Comment by jihunoh-neubla** [comment]
###  **DeepCABAC** 

- deecabac is compression-aware uniform quantization
  - when quantizing weights, the length of the binary encoding of quantized weights is used as a regularization term 
  - so, this produces better compression performance in the following encoding than normal quantization and encoding.
- only layer-wise and symmetric is supported now
- for encoding, we use CABAC which is a lossless binary encoding method

model = yolov3_d53_mstrain-608_273e_coco_optim w/ 80 classes  (original ver.)

quantization & encoding | compression ratio (%) </br> (bin/int8) | zero ratio (%) | mAP
-- | -- | -- | --
fp32 | - | - | 0.309
uniform quant & CABAC | 65.5 | 13 | 0.304
deepcabac quant & CABAC | 56.7 | 19.5 | 0.304

> [detail (uniform quant & CABAC)](https://github.com/NeublaCorp/DeepCABAC/files/8713483/CABAC.yolov3_d53_mstrain-608_273e_coco_optim.q8_lwq_sym_perc_backbone_neck_bridge.compRatio.csv)
> [detail (deepcabac quant & CABAC)](https://github.com/NeublaCorp/DeepCABAC/files/8713496/CABAC.yolov3_d53_mstrain-608_273e_coco_optim.q8_dc_lwq_sym_perc_backbone_neck_bridge.compRatio.csv)


model = yolov3_v23_mstrain-608_273e_coco_optim w/ 6 classes (reduced ver.)

quantization & encoding | compression ratio (%) </br> (bin/int8) | zero ratio (%) | mAP
-- | -- | -- | --
fp32 | - | - | 0.435
uniform quant & CABAC | 62 | 10.5 | 0.431
deepcabac quant & CABAC | 54.5 | 18.1 | 0.432

> [detail (uniform quant & CABAC)](https://github.com/NeublaCorp/DeepCABAC/files/8713485/CABAC.yolov3_v23_mstrain-608_273e_coco_optim.q8_lwq_sym_perc_backbone_neck_bridge_pred.compRatio.csv)
> [detail (deepcabac quant & CABAC)](https://github.com/NeublaCorp/DeepCABAC/files/8713504/CABAC.yolov3_v23_mstrain-608_273e_coco_optim.q8_dc_lwq_sym_perc_backbone_neck_bridge_pred.compRatio.csv)


### Experiment of `interv` param effect in DeepCABAC

- addsteps = 256
- lambda = 0.

interv | 0.7 | 0.6 | 0.5 | 0.4 | 0.3
-- | -- | -- | -- | -- | --
compression ratio (%) | 53.6 | 54.5 | 55.6 | 56.7 | 58
zero ratio (%) | 19.2 | 18.1 | 16.9 | 15.7 | 14.5
mAP | 0.428 | **0.432** | 0.426 | 0.425 | 0.422

### Experiment of `lambda` param effect in DeepCABAC

- addsteps = 256
- interv = 0.6

lambda | 0.0 | 0.001 | 0.01 | 0.1 
-- | -- | -- | -- | -- 
compression ratio (%) | 54.5 | 54.5 | 54.4 |  54.5
zero ratio (%) | 18.1 | 18.1 | 18.3 |  18.3
mAP | 0.432 | 0.432 | 0.432 | 0.407  
 


@deumji-woo @neubla-minwook @wonsubkim-neubla @jong-won-lee

**Comment by jihunoh-neubla** [comment]
### **Asymmetric DeepCABAC quantization**

model = yolov3_v23_mstrain-608_273e_coco_optim w/ 6 classes (reduced ver.)

- addsteps = 256
- interv = 0.6
- lambda = 0.0

quantization & encoding | compression ratio (%) </br> (bin/int8) | zero ratio (%) | mAP
-- | -- | -- | --
fp32 | - | - | 0.435
(Symmetric) deepcabac quant & CABAC | 54.5 | 18.1 | 0.432
Asymmetric deepcabac quant & CABAC | 54.5 | 18.1 | 0.437
